File: exercise/java/week01/exercise/src/main/java/movie/MovieStore.java
package movie;

import java.util.*;

public class MovieStore {
    public HashMap<String, Movie> allMovies;

    public MovieStore() {
        allMovies = new HashMap<>();
    }

    public void addMovie(String id, String title, String director, int totalCopies) {
        if (allMovies.containsKey(id)) {
            System.out.println("Movie already exists! Updating total copies.");
            updateMovieCopies(id, totalCopies);
        } else {
            allMovies.put(id, new Movie(id, title, director, totalCopies));
        }
    }

    public void updateMovieCopies(String id, int newTotalCopies) {
        if (allMovies.containsKey(id)) {
            Movie movie = allMovies.get(id);
            if (newTotalCopies < movie.borrowedCopies) {
                System.out.println("Cannot reduce total copies below borrowed copies.");
            } else {
                movie.totalCopies = newTotalCopies;
            }
        } else {
            System.out.println("Movie not found!");
        }
    }

    public void removeMovie(String id) {
        if (allMovies.containsKey(id)) {
            allMovies.remove(id);
        } else {
            System.out.println("Movie not found!");
        }
    }

    public void borrowMovie(String id) {
        Movie movie = allMovies.get(id);
        if (movie != null) {
            movie.borrowMovie();
        } else {
            System.out.println("Movie not available!");
        }
    }

    public void returnMovie(String id) {
        Movie movie = allMovies.get(id);
        if (movie != null) {
            movie.returnMovie();
        } else {
            System.out.println("Invalid movie ID!");
        }
    }

    public void displayMovies() {
        for (Movie m : allMovies.values()) {
            System.out.println("ID: " + m.movieID + ", Title: " + m.title + ", Director: " + m.director + ", Available Copies: " + (m.totalCopies - m.borrowedCopies));
        }
    }

    public int getTotalMovies() {
        return allMovies.size();
    }

    public List<Movie> findMoviesByTitle(String title) {
        List<Movie> result = new ArrayList<>();
        for (Movie m : allMovies.values()) {
            if (m.title.equalsIgnoreCase(title)) {
                result.add(m);
            }
        }
        return result;
    }

    public List<Movie> getMoviesByDirector(String director) {
        List<Movie> result = new ArrayList<>();
        for (Movie m : allMovies.values()) {
            if (m.director.equals(director)) {
                result.add(m);
            }
        }
        return result;
    }
}

File: exercise/ts/week01-code-analysis/src/movieStore.ts
import {Movie} from "./movie";
import {StoreAccount} from "./storeAccount";

export class MovieStore {
    public allMovies: Map<string, Movie>;
    public sales: StoreAccount;

    constructor() {
        this.allMovies = new Map<string, Movie>();
        this.sales = new StoreAccount();
    }

    buyMovie(customer: string, id: string) {
        const movie = this.allMovies.get(id);
        if (movie != null) {
            if (movie.totalCopies - movie.borrowedCopies > 0) {
                movie.totalCopies--;
                if (movie.canSell()) {
                    this.sales.sell(movie, customer);
                } else {
                    console.log("Movie not for sale");
                }
            } else {
                console.log("All copies are currently borrowed.");
            }
        } else {
            console.log("Movie not available!");
        }
    }

    addMovie(id: string, title: string, director: string, totalCopies: number, unitPrice: number) {
        if (this.allMovies.has(id)) {
            console.log("Movie already exists! Updating total copies.");
            this.updateMovieCopies(id, totalCopies);
        } else {
            this.allMovies.set(id, new Movie(id, title, director, totalCopies, unitPrice));
        }
    }

    updateMovieCopies(id: string, newTotalCopies: number) {
        const movie = this.allMovies.get(id);
        if (movie != null) {
            if (newTotalCopies < movie.borrowedCopies) {
                console.log("Cannot reduce total copies below borrowed copies.");
            } else {
                movie.totalCopies = newTotalCopies;
            }
        } else {
            console.log("Movie not found!");
        }
    }

    removeMovie(id: string) {
        if (this.allMovies.has(id)) {
            this.allMovies.delete(id);
        } else {
            console.log("Movie not found!");
        }
    }

    borrowMovie(id: string) {
        const movie = this.allMovies.get(id);
        if (movie != null) {
            if (movie.totalCopies - movie.borrowedCopies > 0) {
                movie.borrowedCopies++;
            } else {
                console.log("All copies are currently borrowed.");
            }
        } else {
            console.log("Movie not available!");
        }
    }

    returnMovie(id: string) {
        const movie = this.allMovies.get(id);
        if (movie != null) {
            if (movie.borrowedCopies > 0) {
                movie.borrowedCopies--;
            } else {
                console.log("Error: No copies were borrowed.");
            }
        } else {
            console.log("Invalid movie ID!");
        }
    }

    displayMovies() {
        this.allMovies.forEach((m) => {
            console.log(`ID: ${m.movieID}, Title: ${m.title}, Director: ${m.director}, Available Copies: ${m.totalCopies - m.borrowedCopies}`);
        });
    }

    findMoviesByTitle(title: string): Movie[] {
        const result: Movie[] = [];
        this.allMovies.forEach((m) => {
            if (m.title.toLowerCase() === title.toLowerCase()) {
                result.push(m);
            }
        });
        return result;
    }
}